types/wlr_xdg_foreign_v2.c

```diff
@@ -0,0 +1,411 @@
+		struct wlr_xdg_surface *xdg_surface =
+			wlr_xdg_surface_from_wlr_surface(surface);
+		wl_signal_add(&xdg_surface->events.unmap, &exported->xdg_surface_unmap);
+	}
+}
+
+static const struct zxdg_exporter_v2_interface xdg_exporter_impl = {
+	.destroy = xdg_exporter_handle_destroy,
+	.export_toplevel = xdg_exporter_handle_export
+};
+
+static void xdg_exporter_bind(struct wl_client *wl_client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_xdg_foreign_v2 *foreign = data;
+
+	struct wl_resource *exporter_resource =
+		wl_resource_create(wl_client, &zxdg_exporter_v2_interface, version, id);
+	if (exporter_resource == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_resource_set_implementation(exporter_resource, &xdg_exporter_impl,
+			foreign, NULL);
+}
+
+static struct wlr_xdg_foreign_v2 *xdg_foreign_from_importer_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_importer_v2_interface,
+		&xdg_importer_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_importer_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void xdg_imported_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_xdg_imported_v2 *imported = xdg_imported_from_resource(resource);
+
+	struct wlr_xdg_imported_child_v2 *child, *child_tmp;
+	wl_list_for_each_safe(child, child_tmp, &imported->children, link) {
+		struct wlr_xdg_surface *xdg_child =
+			wlr_xdg_surface_from_wlr_surface(child->surface);
+		wlr_xdg_toplevel_set_parent(xdg_child, NULL);
+	}
+
+	wl_list_remove(&imported->exported_destroyed.link);
+	wl_list_remove(&imported->link);
+	free(imported);
+}
+
+static void xdg_imported_handle_exported_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_xdg_imported_v2 *imported =
+		wl_container_of(listener, imported, exported_destroyed);
+	disconnect_imported(imported);
+}
+
+static void xdg_importer_handle_import(struct wl_client *wl_client,
+				struct wl_resource *client_resource,
+				uint32_t id,
+				const char *handle) {
+	struct wlr_xdg_foreign_v2 *foreign =
+		xdg_foreign_from_importer_resource(client_resource);
+
+	struct wlr_xdg_imported_v2 *imported =
+		calloc(1, sizeof(struct wlr_xdg_imported_v2));
+	if (imported == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	imported->exported = wlr_xdg_foreign_registry_find_by_handle(
+		foreign->registry, handle);
+	imported->resource = wl_resource_create(wl_client,
+		&zxdg_imported_v2_interface,
+		wl_resource_get_version(client_resource),
+		id);
+	wl_resource_set_implementation(imported->resource, &xdg_imported_impl,
+			imported, xdg_imported_handle_resource_destroy);
+
+	wl_list_init(&imported->children);
+	wl_list_insert(&foreign->importer.objects, &imported->link);
+
+	if (imported->exported == NULL) {
+		wl_list_init(&imported->exported_destroyed.link);
```

Comment:
I'd prefer to make imported objects without an exported inert, by making the user data pointer NULL and free'ing the `struct wlr_xdg_imported_v2`. This is the standard thing that wlroots does when objects are still alive but can't be used anymore.